### PR TITLE
Add support for Liberica JDK

### PR DESCRIPTION
**Pull request by RazorNd**
**Description:**
Add support for [Liberica JDK](https://bell-sw.com/pages/libericajdk/) distribution.

**Related issue:**
#70 

**Check list:**
- [x] Mark if documentation changes are required.
- [x] Mark if tests were added or updated to cover the changes.

### CODE REVIEW

**Comment by dchuyko** [comment]
Hi. Daniil, a nice PR for Actions. Can we assist somehow? Workflows are awaiting approval for some time, probably should ping maintainers/owners. @chrispat @giltene any advice?

**Comment by maxim-lobanov** [comment]
Hey,
We will take a look at this PR.
cc: @dmitry-shibanov @MaksimZhukov

File: src/distributions/liberica/installer.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+    const availableVersions = availableVersionsRaw.map(item => ({
+      url: item.downloadUrl,
+      version: this.convertVersionToSemver(item)
+    }));
+
+    const satisfiedVersion = availableVersions
+      .filter(item => isVersionSatisfies(range, item.version))
+      .sort((a, b) => -semver.compareBuild(a.version, b.version))
+      .find(first);
```
Comment (maxim-lobanov) [inline src/distributions/liberica/installer.ts:55]: Why do we need this? Can we just take `[0]` element?

File: src/distributions/liberica/installer.ts [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        return { bitness: '64', arch: 'x86' };
+      case 'armv7':
+        return { bitness: '32', arch: 'arm' };
+      case 'aarch64':
+        return { bitness: '64', arch: 'arm' };
+      case 'ppc64le':
+        return { bitness: '64', arch: 'ppc' };
+      default:
+        throw new Error(
+          `Architecture '${this.architecture}' not supported. Supported architecture: ${supportedArchitecture}`
```
Comment (maxim-lobanov) [inline src/distributions/liberica/installer.ts:123]:
```suggestion
          `Architecture '${this.architecture}' is not supported. Supported architectures: ${supportedArchitecture}`
```

File: src/distributions/liberica/installer.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+      case 'win32':
+      case 'cygwin':
+        return 'windows';
+      case 'linux':
+        return 'linux';
+      case 'sunos':
+        return 'solaris';
+      default:
+        throw new Error(
+          `Platform '${platform}' not supported. Supported platform: ${supportedPlatform}`
```
Comment (maxim-lobanov) [inline src/distributions/liberica/installer.ts:141]:
```suggestion
          `Platform '${platform}' is not supported. Supported platforms: ${supportedPlatform}`
```

**Comment by maxim-lobanov** [comment]
@RazorNd, I have left a couple of nitpicks. If you can fix them and pull latest `main` / rebuild code, it would be awesome.
Also, I think @dmitry-shibanov @MaksimZhukov can help with e2e testing of these changes since their team owns this action for now.

File: src/distributions/liberica/installer.ts (hunk 1 above)
Comment (RazorNd) [reply src/distributions/liberica/installer.ts:55]: Good point. I change it.

**Comment by RazorNd** [comment]
> @RazorNd, I have left a couple of nitpicks. If you can fix them and pull latest `main` / rebuild code, it would be awesome. Also, I think @dmitry-shibanov @MaksimZhukov can help with e2e testing of these changes since their team owns this action for now.

Thanks for your corrections. I fixed them. I pulled `main` 6 days ago. This PR is up to date.

**Comment by dchuyko** [comment]
@dmitry-shibanov Are there any issues? Anything that blocks approving the workflows?

**Comment by dmitry-shibanov** [comment]
Hello @RazorNd. Thank you for your pull request. Could you please add e2e tests. I've tested your changes and got error during java installation. I've used such workflow as:
```
name: Test java
on: [push]
jobs:
  run:
    name: Test java ${{ matrix.version }} ${{ matrix.java-package }} ${{ matrix.os }}
    runs-on: ${{ matrix.os }}
    strategy:
      fail-fast: false
      matrix:
        os: [ubuntu-18.04, macos-latest, windows-latest]
        version: [ '8', '11', '13', '15', '16', '13-ea', '15-ea', '16-ea' ]
        java-package: ['jdk', 'jre']
    steps:
    - name: Checkout
      uses: actions/checkout@v2

    - uses: RazorNd/setup-java@liberica
      with:
        java-version: ${{ matrix.version }}
        distribution: liberica
        java-package: ${{ matrix.java-package }}
    - name: Java version
      shell: pwsh
      run: java -version
```

**Comment by dchuyko** [comment]
@RazorNd As far as I can see other recent additions (openj9) included e2e-versions.yml changes.

**Comment by dmitry-shibanov** [comment]
Hello @RazorNd. Thank you for updating branch with e2e tests. I've tested your changes from my side. In general it works as expected, but I have one minor question:
 - Does Liberica jdk provide ea versions ? I've tried to install couple of early access versions, but array of possible versions was empty.

**Comment by RazorNd** [comment]
@dchuyko I was not sure if changing the pipeline for a non-owner is a good idea. But as a last resort I'll revert my changes.

**Comment by RazorNd** [comment]
> Hello @RazorNd. Thank you for updating branch with e2e tests. I've tested your changes from my side. In general it works as expected, but I have one minor question:
> 
> * Does Liberica jdk provide ea versions ? I've tried to install couple of early access versions, but array of possible versions was empty.

I also noticed this yesterday. But if you look in API, Liberica has only 2 early access builds for Java 8. @dchuyko you don't release anymore early access builds? Or am I doing something wrong?

**Comment by dchuyko** [comment]
> * Does Liberica jdk provide ea versions ? I've tried to install couple of early access versions, but array of possible versions was empty.

Currently EA is not among public artifacts (any version).

**Review by Hampedude — APPROVED** [review]
Yeah very good n smart

**Comment by dmitry-shibanov** [comment]
Hello @giltene. If you do not mind, could you please take a look on this pull request ?

**Review by dmitry-shibanov — APPROVED** [review]

File: README.md [hunk 4]
```diff
@@ -57,7 +57,8 @@ Currently, the following distributions are supported:
 | `temurin` | Eclipse Temurin | [Link](https://adoptium.net/) | [Link](https://adoptium.net/about.html)
 | `zulu` | Zulu OpenJDK | [Link](https://www.azul.com/downloads/zulu-community/?package=jdk) | [Link](https://www.azul.com/products/zulu-and-zulu-enterprise/zulu-terms-of-use/) |
 | `adopt` or `adopt-hotspot` | Adopt OpenJDK Hotspot | [Link](https://adoptopenjdk.net/) | [Link](https://adoptopenjdk.net/about.html) |
-| `adopt-openj9` | Adopt OpenJDK OpenJ9 | [Link](https://adoptopenjdk.net/) | [Link](https://adoptopenjdk.net/about.html)
+| `adopt-openj9` | Adopt OpenJDK OpenJ9 | [Link](https://adoptopenjdk.net/) | [Link](https://adoptopenjdk.net/about.html) |
+| `liberica` | Liberica JDK | [Link](https://bell-sw.com/) | [Link](https://bell-sw.com/liberica_eula/) |
```
Comment (MaksimZhukov) [inline README.md:61]: Could you please add and example for Liberica to the [advanced-usage.md](https://github.com/actions/setup-java/blob/main/docs/advanced-usage.md#selecting-a-java-distribution)?

File: .github/workflows/e2e-versions.yml [hunk 5]
```diff
@@ -175,7 +183,7 @@ jobs:
       matrix:
         # Only Zulu provides x86 arch for now and only for windows / ubuntu
         os: [windows-latest, ubuntu-latest]
-        distribution: ['zulu']
+        distribution: ['zulu', 'liberica']
```
Comment (MaksimZhukov) [inline .github/workflows/e2e-versions.yml:186]: Please update the comment above

**Review by MaksimZhukov — COMMENTED** [review]
Hello @RazorNd! Thank you for the contribution! 
Overall looks good to me! I've left a few minor comments

**Comment by RazorNd** [comment]
@MaksimZhukov thanks for you comments. I have resolved all the problems you are reporting.

File: docs/advanced-usage.md [hunk 6]
```diff
@@ -53,6 +53,18 @@ steps:
...
+### Liberica
+```yaml
+steps:
+- uses: actions/checkout@v2
+- uses: actions/setup-java@v2
+  with:
+    distribution: 'liberica'
+    java-version: '11'
+    java-package: jdk # optional (jdk, jre, jdk+fx or jre+fx) - defaults to jdk
+- run: java -cp java HelloWorldApp
```
Comment (MaksimZhukov) [inline docs/advanced-usage.md:65]: Thank you for adding this example! Could you please add a link to it in the [README.md](https://github.com/actions/setup-java#advanced)

Comment (RazorNd) [reply docs/advanced-usage.md:65]: sorry, I missed this point. I'll fix it now.

**Review by MaksimZhukov — APPROVED** [review]

**Comment by dmitry-shibanov** [comment]
Hello @RazorNd. We want to describe our next steps. We will merge your pull request and release a new version next week. We will notify you when we release a new version and update the tag v2.
Thanks for your pull request!

**Pull request merged**